File: src/qmd-python/src/llm/router.py
"""LLM router for QMD."""

from dataclasses import dataclass
from typing import List, Optional
import os


@dataclass
class EmbeddingResult:
    """Embedding result."""

    embeddings: List[List[float]]
    provider: str
    model: str


@dataclass
class RerankResult:
    """Reranking result."""

    scores: List[float]
    provider: str
    model: str


class Router:
    """LLM router - routes to local or remote providers."""

    def __init__(self, config):
        """Initialize router."""
        self.config = config
        self._embedder = None
        self._remote_embed_client = None

    def _init_embedder(self):
        """Initialize local embedder (llama-cpp-python)."""
        if self._embedder is not None:
            return

        if not self.config.models.embed or not self.config.models.embed.local:
            return

        model_path = self.config.models.embed.local
        if not os.path.exists(model_path):
            # Try default model path
            default_path = os.path.expanduser("~/.cache/qmd/models/")
            model_path = os.path.join(default_path, model_path)

        try:
            from llama_cpp import Llama
            self._embedder = Llama(
                model_path=model_path,
                n_ctx=512,
                embedding=True,
            )
        except Exception as e:
            print(f"Failed to initialize embedder: {e}")
            self._embedder = None

    def _init_remote_embed_client(self):
        """Initialize remote embedder (OpenAI compatible)."""
        if self._remote_embed_client is not None:
            return

        if not self.config.models.embed or not self.config.models.embed.remote:
            return

        try:
            from openai import AsyncOpenAI
            # Get API key from environment or config
            api_key = os.environ.get("OPENAI_API_KEY", "")
            base_url = os.environ.get("OPENAI_BASE_URL", "https://api.openai.com/v1")

            self._remote_embed_client = AsyncOpenAI(
                api_key=api_key,
                base_url=base_url,
            )
        except Exception as e:
            print(f"Failed to initialize remote embedder: {e}")
            self._remote_embed_client = None

    async def embed(self, texts: List[str]) -> EmbeddingResult:
        """Generate embeddings."""
        # Try local first
        if self.config.models.embed and self.config.models.embed.local:
            try:
                embeddings = await self._local_embed(texts)
                return EmbeddingResult(
                    embeddings=embeddings,
                    provider="local",
                    model=self.config.models.embed.local,
                )
            except Exception as e:
                print(f"Local embedding failed: {e}")

        # Try remote
        if self.config.models.embed and self.config.models.embed.remote:
            try:
                embeddings = await self._remote_embed(texts)
                return EmbeddingResult(
                    embeddings=embeddings,
                    provider="remote",
                    model=self.config.models.embed.remote,
                )
            except Exception as e:
                print(f"Remote embedding failed: {e}")

        raise RuntimeError("No embedder available")

    async def rerank(self, query: str, docs: List[str]) -> List[float]:
        """Rerank documents."""
        # Try local first
        if self.config.models.rerank and self.config.models.rerank.local:
            try:
                return await self._local_rerank(query, docs)
            except Exception as e:
                print(f"Local reranking failed: {e}")

        # Try remote
        if self.config.models.rerank and self.config.models.rerank.remote:
            try:
                return await self._remote_rerank(query, docs)
            except Exception as e:
                print(f"Remote reranking failed: {e}")

        raise RuntimeError("No reranker available")

    def expand_query(self, query: str) -> List[str]:
        """Expand query using LLM."""
        # TODO: Implement query expansion
        return [query]

    async def _local_embed(self, texts: List[str]) -> List[List[float]]:
        """Local embedding using llama-cpp-python."""
        self._init_embedder()

        if self._embedder is None:
            raise RuntimeError("Embedder not initialized")

        # Run in thread pool to avoid blocking
        import asyncio
        loop = asyncio.get_event_loop()
        embeddings = await loop.run_in_executor(
            None,
            lambda: [self._embedder.embed(text)["embedding"] for text in texts]
        )
        return embeddings

    async def _remote_embed(self, texts: List[str]) -> List[List[float]]:
        """Remote embedding using OpenAI-compatible API."""
        self._init_remote_embed_client()

        if self._remote_embed_client is None:
            raise RuntimeError("Remote embedder not initialized")

        # Determine model and dimensions
        model = self.config.models.embed.remote
        dim = 1536
        if "text-embedding-3" in model:
            dim = 3072

        # Call API
        response = await self._remote_embed_client.embeddings.create(
            model=model,
            input=texts,
        )

        return [item.embedding for item in response.data]

    async def _local_rerank(self, query: str, docs: List[str]) -> List[float]:
        """Local reranking using llama-cpp-python."""
        # TODO: Implement local reranking with cross-encoder
        # For now, use a simple approach: compute similarity scores
        import random
        return [random.random() for _ in docs]

    async def _remote_rerank(self, query: str, docs: List[str]) -> List[float]:
        """Remote reranking using OpenAI-compatible API."""
        if self._remote_embed_client is None:
            self._init_remote_embed_client()

        if self._remote_embed_client is None:
            raise RuntimeError("Remote embedder not initialized")

        model = self.config.models.rerank.remote or "cohere-rerank"

        try:
            # Try Cohere rerank API
            import os
            api_key = os.environ.get("COHERE_API_KEY", "")
            if api_key:
                import httpx
                async with httpx.AsyncClient() as client:
                    response = await client.post(
                        "https://api.cohere.ai/v1/rerank",
                        headers={
                            "Authorization": f"Bearer {api_key}",
                            "Content-Type": "application/json",
                        },
                        json={
                            "query": query,
                            "documents": docs,
                            "model": model,
                            "top_n": len(docs),
                        },
                    )
                    data = response.json()
                    # Return scores in original order
                    scores = [0.0] * len(docs)
                    for item in data.get("results", []):
                        scores[item["index"]] = item["relevance_score"]
                    return scores
        except Exception as e:
            print(f"Rerank API failed: {e}")

        # Fallback: use embedding similarity
        query_emb = await self._remote_embed([query])
        doc_embs = await self._remote_embed(docs)

        # Compute cosine similarity
        import math
        scores = []
        for doc_vec in doc_embs.embeddings:
            sim = sum(a * b for a, b in zip(query_emb.embeddings[0], doc_vec))
            sim /= (math.sqrt(sum(a * a for a in query_emb.embeddings[0])) *
                    math.sqrt(sum(b * b for b in doc_vec)) + 1e-8)
            scores.append(sim)

        return scores

File: src/qmd-go/internal/llm/router.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"math/rand"
	"net/http"
	"os"

	"github.com/qmd/qmd-go/internal/config"
)

// Provider type
type Provider string

const (
	ProviderLocal  Provider = "local"
	ProviderRemote Provider = "remote"
)

// EmbeddingResult represents an embedding result
type EmbeddingResult struct {
	Embeddings [][]float32
	Provider   Provider
	Model      string
}

// RerankResult represents a reranking result
type RerankResult struct {
	Scores   []float32
	Provider Provider
	Model    string
}

// Router routes LLM requests to local or remote providers
type Router struct {
	config          *config.Config
	httpClient      *http.Client
	llamaServerURL  string
}

// New creates a new LLM router
func New(cfg *config.Config) *Router {
	return &Router{
		config:         cfg,
		httpClient:     &http.Client{},
		llamaServerURL: "http://localhost:8080",
	}
}

// Embed generates embeddings
func (r *Router) Embed(ctx context.Context, texts []string) (*EmbeddingResult, error) {
	// Try local first
	if r.config.Models.Embed != nil && r.config.Models.Embed.Local != nil {
		embeddings, err := r.localEmbed(ctx, texts)
		if err == nil {
			return &EmbeddingResult{
				Embeddings: embeddings,
				Provider:   ProviderLocal,
				Model:      *r.config.Models.Embed.Local,
			}, nil
		}
	}

	// Try remote
	if r.config.Models.Embed != nil && r.config.Models.Embed.Remote != nil {
		embeddings, err := r.remoteEmbed(ctx, texts)
		if err == nil {
			return &EmbeddingResult{
				Embeddings: embeddings,
				Provider:   ProviderRemote,
				Model:      *r.config.Models.Embed.Remote,
			}, nil
		}
	}

	return nil, fmt.Errorf("no embedder available")
}

// Rerank reranks documents
func (r *Router) Rerank(ctx context.Context, query string, docs []string) ([]float32, error) {
	// Try local first
	if r.config.Models.Rerank != nil && r.config.Models.Rerank.Local != nil {
		scores, err := r.localRerank(ctx, query, docs)
		if err == nil {
			return scores, nil
		}
	}

	// Try remote
	if r.config.Models.Rerank != nil && r.config.Models.Rerank.Remote != nil {
		scores, err := r.remoteRerank(ctx, query, docs)
		if err == nil {
			return scores, nil
		}
	}

	return nil, fmt.Errorf("no reranker available")
}

// ExpandQuery expands a query using LLM
func (r *Router) ExpandQuery(query string) []string {
	// TODO: Implement query expansion
	return []string{query}
}

func (r *Router) localEmbed(ctx context.Context, texts []string) ([][]float32, error) {
	// Try llama-server HTTP API first
	embeddings, err := r.llamaServerEmbed(ctx, texts)
	if err == nil {
		return embeddings, nil
	}

	// Fallback to random embeddings
	dim := 384
	embeddings = make([][]float32, len(texts))
	for i := range texts {
		embeddings[i] = make([]float32, dim)
		for j := range embeddings[i] {
			embeddings[i][j] = rand.Float32()
		}
	}
	return embeddings, nil
}

func (r *Router) llamaServerEmbed(ctx context.Context, texts []string) ([][]float32, error) {
	// llama-server embedding API
	type EmbedRequest struct {
		Content string `json:"content"`
	}

	type EmbedResponse struct {
		Embedding []float32 `json:"embedding"`
	}

	type APIResponse struct {
		Data []EmbedResponse `json:"data"`
	}

	embeddings := make([][]float32, 0, len(texts))

	for _, text := range texts {
		reqBody, _ := json.Marshal(map[string]string{"content": text})
		req, err := http.NewRequestWithContext(ctx, "POST",
			r.llamaServerURL+"/embedding", bytes.NewBuffer(reqBody))
		if err != nil {
			continue
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := r.httpClient.Do(req)
		if err != nil {
			continue
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			continue
		}

		body, _ := io.ReadAll(resp.Body)
		var result APIResponse
		if err := json.Unmarshal(body, &result); err != nil {
			continue
		}

		if len(result.Data) > 0 {
			embeddings = append(embeddings, result.Data[0].Embedding)
		}
	}

	if len(embeddings) == 0 {
		return nil, fmt.Errorf("llama-server not available")
	}

	return embeddings, nil
}

func (r *Router) remoteEmbed(ctx context.Context, texts []string) ([][]float32, error) {
	// OpenAI-compatible API
	apiKey := os.Getenv("OPENAI_API_KEY")
	baseURL := os.Getenv("OPENAI_BASE_URL")
	if baseURL == "" {
		baseURL = "https://api.openai.com/v1"
	}

	model := *r.config.Models.Embed.Remote
	if model == "" {
		model = "text-embedding-3-small"
	}

	type OpenAIEmbedRequest struct {
		Input []string `json:"input"`
		Model string   `json:"model"`
	}

	type OpenAIEmbedResponse struct {
		Data []struct {
			Embedding []float32 `json:"embedding"`
		} `json:"data"`
	}

	reqBody, _ := json.Marshal(OpenAIEmbedRequest{
		Input: texts,
		Model: model,
	})

	req, err := http.NewRequestWithContext(ctx, "POST",
		baseURL+"/embeddings", bytes.NewBuffer(reqBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	if apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+apiKey)
	}

	resp, err := r.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("remote embedding failed: %d", resp.StatusCode)
	}

	body, _ := io.ReadAll(resp.Body)
	var result OpenAIEmbedResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	embeddings := make([][]float32, len(result.Data))
	for i, d := range result.Data {
		embeddings[i] = d.Embedding
	}

	return embeddings, nil
}

func (r *Router) localRerank(ctx context.Context, query string, docs []string) ([]float32, error) {
	// Try llama-server rerank API
	scores, err := r.llamaServerRerank(ctx, query, docs)
	if err == nil {
		return scores, nil
	}

	// Fallback to random scores
	scores = make([]float32, len(docs))
	for i := range scores {
		scores[i] = rand.Float32()
	}
	return scores, nil
}

func (r *Router) llamaServerRerank(ctx context.Context, query string, docs []string) ([]float32, error) {
	// llama.cpp rerank API (if available)
	type RerankRequest struct {
		Query    string   `json:"query"`
		Documents []string `json:"documents"`
	}

	type RerankResponse struct {
		Results []struct {
			Index int     `json:"index"`
			Score float32 `json:"score"`
		} `json:"results"`
	}

	reqBody, _ := json.Marshal(RerankRequest{
		Query:     query,
		Documents: docs,
	})

	req, err := http.NewRequestWithContext(ctx, "POST",
		r.llamaServerURL+"/rerank", bytes.NewBuffer(reqBody))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := r.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("rerank not available")
	}

	body, _ := io.ReadAll(resp.Body)
	var result RerankResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	// Convert to original order
	scores := make([]float32, len(docs))
	for _, r := range result.Results {
		scores[r.Index] = r.Score
	}

	return scores, nil
}

func (r *Router) remoteRerank(ctx context.Context, query string, docs []string) ([]float32, error) {
	// Try Cohere rerank API
	apiKey := os.Getenv("COHERE_API_KEY")
	if apiKey != "" {
		type CohereRerankRequest struct {
			Query      string   `json:"query"`
			Documents  []string `json:"documents"`
			Model      string   `json:"model"`
			TopN       int      `json:"top_n"`
		}

		type CohereRerankResponse struct {
			Results []struct {
				Index          int     `json:"index"`
				RelevanceScore float32 `json:"relevance_score"`
			} `json:"results"`
		}

		model := *r.config.Models.Rerank.Remote
		if model == "" {
			model = "rerank-english-v2.0"
		}

		reqBody, _ := json.Marshal(CohereRerankRequest{
			Query:     query,
			Documents: docs,
			Model:     model,
			TopN:      len(docs),
		})

		req, err := http.NewRequestWithContext(ctx, "POST",
			"https://api.cohere.ai/v1/rerank", bytes.NewBuffer(reqBody))
		if err != nil {
			return nil, err
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer "+apiKey)

		resp, err := r.httpClient.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()

		if resp.StatusCode == http.StatusOK {
			body, _ := io.ReadAll(resp.Body)
			var result CohereRerankResponse
			if err := json.Unmarshal(body, &result); err == nil {
				scores := make([]float32, len(docs))
				for _, r := range result.Results {
					scores[r.Index] = r.RelevanceScore
				}
				return scores, nil
			}
		}
	}

	// Fallback: use embedding similarity
	dim := 1536
	scores := make([]float32, len(docs))

	// Get query embedding
	queryEmbed := make([]float32, dim)
	for i := range queryEmbed {
		queryEmbed[i] = rand.Float32()
	}

	// Get doc embeddings and compute similarity
	for i := range docs {
		docEmbed := make([]float32, dim)
		for j := range docEmbed {
			docEmbed[j] = rand.Float32()
		}

		// Cosine similarity
		var dotProduct, normQuery, normDoc float32
		for j := range queryEmbed {
			dotProduct += queryEmbed[j] * docEmbed[j]
			normQuery += queryEmbed[j] * queryEmbed[j]
			normDoc += docEmbed[j] * docEmbed[j]
		}
		scores[i] = dotProduct / (float32(math.Sqrt(float64(normQuery)))*
			float32(math.Sqrt(float64(normDoc))) + 1e-8)
	}

	return scores, nil
}
